serve/mlc_serve/model/paged_cache_model.py

```diff
@@ -125,25 +178,38 @@ def _is_safe_to_sample(prob_like):
         logits_random.div_(t.unsqueeze(dim=1))
 
     if do_top_p or do_top_k:
-        logits = _apply_top_p_top_k(logits_random, top_ps, top_ks)
+        logits_random = _apply_top_p_top_k(logits_random, top_ps, top_ks)
 
     probs = torch.softmax(logits_random, dim=-1)
 
     if check_safety and not _is_safe_to_sample(probs):
         return None
 
-    res_random = torch.multinomial(probs, 1, True).cpu().numpy()[:, 0]
+    res_random = torch.multinomial(probs, 1, True)[:, 0]
+
+    logprob_infos_random = fetch_raw_logprob_infos(
+        logits_random,
+        res_random,
+        sampling_params[mask_random],
+    )
 
+    res_random = res_random.cpu().numpy()
+    # Case when there's only random sampling
     if logits_random.shape[0] == num_seq:
-        return res_random
+        return res_random, logprob_infos_random
 
     res = np.empty((num_seq,), dtype=np.int32)
     res[mask_random] = res_random
 
+    logprob_infos: List[Optional[RawLogprobsInfo]] = [None] * num_seq
+    logprob_infos[mask_random] = logprob_infos_random
```

Comment:
Have you tested this code path? I don't think we can do such indexing into a python list.

Reply:
Hello @zxybazh! Could you use diff below to fix the bug and check full pipeline with it?
```
diff --git a/serve/mlc_serve/model/paged_cache_model.py b/serve/mlc_serve/model/paged_cache_model.py
index c663dec..5cfbf9d 100644
--- a/serve/mlc_serve/model/paged_cache_model.py
+++ b/serve/mlc_serve/model/paged_cache_model.py
@@ -100,6 +100,16 @@ def _apply_top_p_top_k(logits, top_ps, top_ks):
     return logits
 
 
+def update_masked_list(input_list, mask, update):
+    j = 0
+    for i in range(len(mask)):
+        if mask[i]:
+            input_list[i] = update[j]
+            j = j + 1
+
+    return input_list
+
+
 def sample(
     logits: Union[tvm.nd.NDArray, torch.Tensor],
     sampling_params: List[SamplingParams],
@@ -202,12 +212,12 @@ def sample(
     res[mask_random] = res_random
 
     logprob_infos: List[Optional[RawLogprobsInfo]] = [None] * num_seq
-    logprob_infos[mask_random] = logprob_infos_random
+    logprob_infos = update_masked_list(logprob_infos, mask_random, logprob_infos_random)
 
     if logits_greedy.shape[0] > 0:
         res[mask_greedy] = res_greedy
 
-        logprob_infos[mask_greedy] = logprob_infos_greedy
+        logprob_infos = update_masked_list(logprob_infos, mask_greedy, logprob_infos_greedy)
 
     return res, logprob_infos
 
```